### PR TITLE
Add charts to store and node pages.

### PR DESCRIPTION
**Pull request by BramGruneir**
Adds:
- Stats and Charts on the store/:store_id page
- Stats and Charts on the node/:node_id pages
- Removes dead links from the header.

This is still rough, but I wanted to get something in for the demo tomorrow.  Goal for this PR was purely functionality.

Note that I couldn't get the byte charts working as there is a bug in the line chart code.  I'll try to fix that tomorrow.

Future work to be done:
- There's going to be lots of work refactoring the code and adding all the missing comments.
- I'd like to batch the queries together instead of having each one be a different network call,
- Lots of design work to make the pages look nicer.
- More work on the charts.
- I'm going to add the statistics table to the node and store pages.  This should be easy to do, I'll try to get that in tomorrow as well. --- This is now included in this PR.

Store Page 1/2:
screen shot 2015-05-28 at 03 32 42

Store Page 2/2:
screen shot 2015-05-28 at 03 32 58

Node Page 1/2:
screen shot 2015-05-28 at 03 33 27

Node Page 2/2:
screen shot 2015-05-28 at 03 33 36

### CODE REVIEW

**Comment by petermattis** [comment]
The "Successful Calls" and "Error Calls" graphs are not too useful. I'd dump them and only keep the associated "Rate" graphs.

**Comment by BramGruneir** [comment]
I can drop those two.  How's it looking otherwise?

**Comment by petermattis** [comment]
It look functional for the demo. We're going to get design help to make it pretty in the coming weeks.

I noticed the tooltips on your graphs are nowhere near where the graph is. Bug?

**Comment by BramGruneir** [comment]
Great.
And yeah, that's another bug, they should appear near the mouse.  I'm going to see if I can get that fixed as well.

**Comment by BramGruneir** [comment]
Removed the success and error chart from the node detail page.
Added all nodes and stores statistics to the listing pages.

Note that all values are polled every 10s from the webpage.

screen shot 2015-05-28 at 11 54 16
screen shot 2015-05-28 at 11 54 25

File: resource/ui/ts/models/node_status.ts [hunk 1]
```diff
@@ -141,10 +139,70 @@ module Models {
...
+                        live_count: 0,
+                        key_count: 0,
+                        val_count: 0,
+                        intent_count: 0,
+                        sys_bytes: 0,
+                        sys_count: 0
+                    }
+                };
+
+                var nodeId: string;
```
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: do you really need this declaration outside the `for (...) ` below/

File: resource/ui/ts/models/node_status.ts [hunk 2]
```diff
@@ -141,10 +139,70 @@ module Models {
                         m("tr", [m("td", "Ranges:"), m("td", node.range_count)]),
                         m("tr", [m("td", "Leader Ranges:"), m("td", node.leader_range_count)]),
                         m("tr", [m("td", "Available Ranges:"), m("td", node.available_range_count)]),
-                        m("tr", [m("td", "Availablility:"), m("td", this._availability(nodeId))]),
+                        m("tr", [m("td", "Availablility:"), m("td", Nodes._availability(node))]),
```
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: wouldn't `this` still work? more generally: what is the value of `this` in a typescript static method?

File: resource/ui/ts/models/node_status.ts [hunk 3]
```diff
@@ -141,10 +139,70 @@ module Models {
...
+                for (nodeId in this.statuses()) {
+                    var nodeStatus = this.statuses()[nodeId];
+                    status.range_count += nodeStatus.range_count;
+                    status.leader_range_count += nodeStatus.leader_range_count;
+                    status.replicated_range_count += nodeStatus.replicated_range_count;
+                    status.available_range_count += nodeStatus.available_range_count;
+                    if (nodeStatus.updated_at > status.updated_at) {
+                        status.updated_at = nodeStatus.updated_at;
+                    }
+                    status.stats.live_bytes += nodeStatus.stats.live_bytes;
```
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: this field-by-field addition seems fraught with danger. is it possible to implement the addition operator in TS? might be good to do in a future PR.

File: resource/ui/ts/models/stats.ts [hunk 4]
```diff
@@ -22,5 +22,61 @@ module Models {
             sys_count: number;
             last_update_nanos: number;
         }
+
+        // This function was adapted from
+        // https://stackoverflow.com/questions/10420352/converting-file-size-in-bytes-to-human-readable
+        export function FormatBytes(bytes: number): string {
+            var thresh = 1024;
```
Comment (tamird) [inline resource/ui/ts/models/stats.ts]: does ts have a notion of constants?

Comment (tamird) [inline resource/ui/ts/models/stats.ts]: also perhaps a better name is `kilo`?

Comment (tamird) [inline resource/ui/ts/models/stats.ts]: `kibi`, ratehr

File: resource/ui/ts/models/node_status.ts (hunk 1 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: I tried putting it in the for with a 
for (var nodeID in ...
But it kept failing.  I'll try it again.

File: resource/ui/ts/models/stats.ts [hunk 5]
```diff
@@ -22,5 +22,61 @@ module Models {
...
+            if (Math.abs(bytes) < thresh) {
+                return bytes + ' B';
+            }
+            var units = ['KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'];
```
Comment (tamird) [inline resource/ui/ts/models/stats.ts]: nit: you can simplify this function by prepending `'B'` to this array (and making it a constant) and indexing into it by `log2(bytes)/10`.

[log2](https://github.com/Microsoft/TypeScript/blob/master/src/lib/es6.d.ts#L541

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: Static method exists on the class itself, so you have to reference it using the classname.
`this` in a static method should refer to the class itself, not the instance (since you wouldn't have one).  But with JS, the context can get mixed up quickly.  See the func.call(instance) that I have to do in the refresher.

File: resource/ui/ts/models/store_status.ts [hunk 6]
```diff
@@ -136,10 +134,69 @@ module Models {
...
+                    status.leader_range_count += storeStatus.leader_range_count;
+                    status.replicated_range_count += storeStatus.replicated_range_count;
+                    status.available_range_count += storeStatus.available_range_count;
+                    if (storeStatus.updated_at > status.updated_at) {
+                        status.updated_at = storeStatus.updated_at;
+                    }
+                    status.stats.live_bytes += storeStatus.stats.live_bytes;
+                    status.stats.key_bytes += storeStatus.stats.key_bytes;
+                    status.stats.val_bytes += storeStatus.stats.val_bytes;
+                    status.stats.intent_bytes += storeStatus.stats.intent_bytes;
```
Comment (tamird) [inline resource/ui/ts/models/store_status.ts]: This looks like the same code from node_status. Add a TODO to extract?

File: resource/ui/ts/models/node_status.ts (hunk 3 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: Yeah, this is rather ugly.  I'll add a todo about a refactoring here.

File: resource/ui/ts/pages/nodes.ts [hunk 7]
```diff
@@ -13,14 +15,66 @@ module AdminViews {
...
+      private _nodeId: string;
+      private _refreshFunctions: RefreshFunction[]
+
+      private _refresh():void {
+        for (var i = 0; i < this._refreshFunctions.length; i++) {
+          this._refreshFunctions[i].f.call(this._refreshFunctions[i].o);
+        }
+      }
+
+      private static _queryManagerBuilder(nodeId: string, agg: Models.Metrics.QueryAggregator, source:string): Models.Metrics.QueryManager {
```
Comment (tamird) [inline resource/ui/ts/pages/nodes.ts]: i dont meaningfully understand this code, so not reviewing it in detail. nit: space after `source:`

File: resource/ui/ts/models/stats.ts (hunk 4 above)
Comment (BramGruneir) [inline resource/ui/ts/models/stats.ts]: TS doesn't seem to have constants. I'll pull it out of the function at least.  And yes, I'll change the name to kibi. :)

File: resource/ui/ts/typings/mithriljs/mithril.d.ts [hunk 8]
```diff
@@ -122,8 +122,8 @@ declare module _mithril {
 		toJSON(): T;
 	}
 
-	interface MithrilRoutes<T extends MithrilController> {
```
Comment (tamird) [inline resource/ui/ts/typings/mithriljs/mithril.d.ts]: why these changes?

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: Whoa, that is bonkers.

> this in a static method should refer to the class itself

What are describing here? Not ts/js behaviour, I take it?

File: resource/ui/ts/pages/nodes.ts (hunk 7 above)
Comment (BramGruneir) [inline resource/ui/ts/pages/nodes.ts]: This is the fancy JS I had to come up with to get multiple charts working in a sane manner.  
The builder creates instances of the queryManager (each one essentially represents a single chart on a page at this point.)
The _refresh function is where all the requery logic sits, to update the data every interval.

File: resource/ui/ts/typings/mithriljs/mithril.d.ts (hunk 8 above)
Comment (BramGruneir) [inline resource/ui/ts/typings/mithriljs/mithril.d.ts]: The typescript typings were incorrect.  Matt was the one who actually did the work on this.  If we don't do this, the transpiler fails when it tries to sort out the routing table in app.ts.

Comment (tamird) [inline resource/ui/ts/typings/mithriljs/mithril.d.ts]: K.

File: resource/ui/ts/models/node_status.ts (hunk 1 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: Found the issue.  Fixed.

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: So this was interesting so I wanted to test it out.
When in a static method, the `this` refers to the class (or to be more exact, the function that is the definition of the class)
`
this
Stores()
`
referes to:
`
 var Stores = (function () {
            function Stores() {
                this._data = m.prop({});
                this.desc = m.prop({});
                this.statuses = m.prop({});
            }
            Stores.prototype.Query = function () {
  ...`

Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: So in the end, when in a static function, don't use the `this`.  Ever.

File: resource/ui/ts/models/node_status.ts (hunk 3 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: done.

File: resource/ui/ts/models/stats.ts (hunk 5 above)
Comment (BramGruneir) [inline resource/ui/ts/models/stats.ts]: Converted to constants.  I'll update the function using log2 in another PR.

File: resource/ui/ts/models/store_status.ts (hunk 6 above)
Comment (BramGruneir) [inline resource/ui/ts/models/store_status.ts]: done.

**Comment by BramGruneir** [comment]
updated, @tamird PTAL

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: wait, what's at the end of those ellipsis? if `this` is just a value, why not use it?

File: resource/ui/ts/models/stats.ts [hunk 9]
```diff
@@ -22,5 +22,61 @@ module Models {
             sys_count: number;
             last_update_nanos: number;
         }
+
+        // This function was adapted from
+        // https://stackoverflow.com/questions/10420352/converting-file-size-in-bytes-to-human-readable
+        var kibi = 1024
```
Comment (tamird) [inline resource/ui/ts/models/stats.ts]: did you see my other comment about using log?

**Comment by tamird** [comment]
I don't want to block you, so LGTM. It'd be nice to address the line of discussion on the use of `this` and to simplify `FormatBytes` using `log2`.

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: That's just from the generated code.  I just cut out a section of it and added the ellipsis myself.
this = Stores() -- when you're inside a static function.
When you're in an inside a non-static function, `this` points to the instance of the class.  You cannot access any static functions from the `this` pointer if you're inside an instance. since the functions are not on the instance.

Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: right, but in this case you're inside a static method, so `this` should be the class instance?

**Comment by BramGruneir** [comment]
I think I've addressed the `this` usage sufficiently.  
I'll convert FormatBytes to use log2 in another PR.

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: oh, nm, this is not a static method. Ignore me and carry on! though, something akin to ruby's  `self.class` would be neat.

File: resource/ui/ts/models/stats.ts (hunk 9 above)
Comment (BramGruneir) [inline resource/ui/ts/models/stats.ts]: yep, responded to it.

File: resource/ui/ts/models/node_status.ts [hunk 10]
```diff
@@ -78,38 +78,34 @@ module Models {
 
             private _updateDescriptions(): void {
                 this.desc(<NodeDescriptionMap> {});
-                var nodeId: string;
-                for (nodeId in this._data()) {
+                for (var nodeId in this._data()) {
```
Comment (tamird) [inline resource/ui/ts/models/node_status.ts]: neat

File: resource/ui/ts/models/node_status.ts (hunk 2 above)
Comment (BramGruneir) [inline resource/ui/ts/models/node_status.ts]: Agreed, self.class would be great.
And even if we were in a static function, the this doesn't seem to give you access to the other static elements, it's just a function representing the class.  Very strange.  I tried chaining things off of it, and it wouldn't let me in the debugger.
